### ISSUE TITLE
No point tracking Cloud IPs

### ISSUE THREAD

**Issue by mikeruss1** [CIDRAM/CIDRAM#199]
on a connected suggestion to the last one, is there any point tracking IPs from a cloud service? Just filling up tracking files and taking up processing resource?

**Comment by Maikuolan** [comment]
If they're blocked by normal signature files, not really. The main benefit of IP tracking (in the CIDRAM context) is being able to gradually scale up CIDRAM's response to unwanted traffic in cases where said traffic may be being flagged as unwanted for reasons other than the IP address itself (e.g., requests blocked by modules due to user agents of known malicious entities, detected hack attempts, other malicious behaviour, etc). When we see a certain amount of unwanted traffic from a particular address, we can begin outright blocking that address for any subsequent requests under the assumption that any subsequent traffic from that address will be unwanted, regardless of the results of any further processing. On the other hand, if that IP address is already blocked by normal signature files anyway, then there isn't really anything to "scale up" in the first place, since the address itself is already outright blocked.

**Comment by Maikuolan** [comment]
The other benefit is being able to change the response once an address becomes "banned" (i.e., the IP tracking threshold is reached or exceeded), potentially slightly reducing bandwidth usage and such. That's only relevant though for those that want to utilise that particular option (`ban_override`).

**Comment by mikeruss1** [comment]
my idea was that a cloud service bot just rotates through thousands of IP addresses. We block the first one, then we never see it again cos he is making his way through all the others, so no value in tracking it?

**Comment by Maikuolan** [comment]
Ah, okay. Well.. That's a kinda sorta there, I think. Hindsight is always 20/20, but until we actually know, since we generally won't actually know whether we'll ever see it again most of the time, IP tracking provides CIDRAM a way to recognise when it's seeing the same addresses again and again (or, when not). If we already know in advance that we're going to be seeing the same addresses again and again, then at that point, it's better to just outright block it via normal signature files, via a normal CIDR signature, anyway. If we don't end up ever seeing a particular address ever again, then that's fine too, and the IP tracking entries for it will eventually expire and disappear anyway, and if we do, then it'll follow through the usual process until eventually reaching "banned" status. Of course, if foresight was also always 20/20, we most likely wouldn't need an IP tracking feature to begin with, and I'd then be able to consider just stripping it out entirely for the next major version.

If there are particular bots that we're definitely 100% sure won't ever bother using the same IP addresses more than once, and if we're also definitely 100% sure that no *other* bots will be using those same IP addresses either (since IP tracking tracks the IP address, rather than the particular bots in question, thus multiple, different unwanted bots operating from the same IP address will all contribute towards the infraction count for that IP address), then we could, if we want, opt out of IP tracking for that particular bot.

Though, if we were to start opting out for specific bots, we'd still need to be identifying which bots we're opting out for, and which bots we're choosing to remain opted in for (a little more work, unless we just avoid responsibility for it and push that work out to the end-users, of course), which then raises the question of whether it's worth the extra work to begin with, and if not, then whether it's more worthwhile to just disable IP tracking entirely for everything (not trying to be dramatic; just exploring upon where our priorities are, or should be, in regards to our own personal time, efforts, etc), or alternatively, whether to just keep it enabled entirely for everything. Though TBH, I don't really know where we should sit in that regard; Just thinking out loud. `¯\_(ツ)_/¯`

If anyone wants a hand disabling IP tracking for specific bots, signatures, etc, I'd be more than happy to give them a hand. I'm not really convinced that I should be taking a hard stance on the matter either way though, and would be more inclined to leave it as something that end-users can decide for themselves.

**Comment by mikeruss1** [comment]
suspect the benefits of this one are marginal. If its easy and you have the time, then an option to not track IPs resulting from 
cloud service blocks would be my choice. The IP in this case is essentially a random number as it doesnt represent a physical location. I dont see the UA that is causing it as material.

**Comment by 737simpilot** [comment]
This is solely up to the user to disable cloud blocking of their IPs. There are MANY cloud providers that sit in front of services that CAN be used to access your website resources. To out right get rid of that is stupid. And by definition, "cloud" means CloudFlare, Stack Path, Akamai, (another I can't think of the name) moving to Azure, Digital Ocean, Linode, you name it. ALL encompass the "cloud." It would be extremely careless to blanket omit these IPs. And there is no reason once so ever for a server to connect to a server. You only do that on a temp basis if you're running some online website speed check or something. And at that it should be at a temporally basis. Once you check your webpage speed, TLS cert or what ever the hell you need server to server communications, you reban the IP. The proper way for that would be a combination UA/IP/CIDR combination bypass. You DON'T want to blanket allow the UA and that's it. 

Question: Does legit traffic ever come from a cloud IP? No? Then why blanket ditch the cloud IPs from being blocked? Where's the logic?

And the word is cause, not "cos."

Now I gotta ditch Coinbase because these incompetent losers ditched Authy 2FA for CRAP! To say I'm pissed right now is a gross universal understatement. Then I read this.

**Comment by mikeruss1** [comment]
I prefer cos in that sentence

**Comment by 737simpilot** [comment]
LOL What ever floats your boat.

**Comment by Maikuolan** [comment]
Instead of bothering with new configuration directives (I mean.. there's quite a lot of them already, after all..), I'm thinking maybe something like adding a new field to the auxiliary rules, to enable tracking to be optionally toggled on or off. That way, enabling/disabling tracking could be handled as a rule, and would have the same fluidity that auxiliary rules provide (e.g., being able to add conditions based on URLs, based on IPs, based on block reason, etc), as well as avoiding the need for new setups to go through additional configuration directives to be customised according to the needs of the setup. On the other hand.. Adding new configuration directives, while it may immediately sound like an easier solution, would likely end up feeling much more strict and rigid in the long term, since each condition or permutation affecting weather to turn tracking on or off would likely each requires its own configuration directive, further bloating the number that already exist. Thoughts?

**Comment by mikeruss1** [comment]
Agreed.
Reason = Cloud Service (or hybrid) - option = turn off tracking.
the event would need to be logged and reported to user as normal

the result of course would be that multiple blocks on the same IP would be logged and user receives block page. On my site its generally either 1 or 2 blocks on an IP, or the bot users multiple IP addresses, so I am OK with that outcome.

**Comment by Maikuolan** [comment]
Er.. Okay. I feel a little dumb now. I just realised, the exact thing I suggested adding two days ago here (reply above), I'd actually already implemented last year. I guess I just completely forgot about it at the time, when suggesting to add it again here. Not sure what it says about me, that I'm forgetting things that I'd written and implemented myself. Anyway.. My bad. <.<

A slight difference, however, between how I'd already implemented it and the manner in which I'd suggested to implement it above, is that the existing implementation is a simple switch under the "other options and special flags" header, whereas my suggestion above was to implement a toggle. With that difference in mind.. For it to be useful, I'll probably need to add a new option to the action list (alongside the others like block, greylist, whitelist, bypass, run, etc), to be able to utilise the switch without performing any of the existing actions (since we're not wanting to perform additional actions). That'll help with some other suggestions others have made recently, too.

I'll reply again a little later, after adding the new action.

**Comment by mikeruss1** [comment]
very efficient, waves magic wand and its already done

**Comment by Maikuolan** [comment]
Latest update to the auxiliary rules system has just been pushed. :-)

If you check it out, there are some new things you'll notice:

- IP tracking can be forcibly enabled or forcibly disabled for when a rule is triggered (now properly toggleable, instead of just a linear switch like before).
- In the actions list, there's a new action: *"If the following conditions are met, __profile__ the request"*. Unlike the other actions, selecting that action will elicit no apparent difference to the end-user for whether the rule is or isn't triggered, having no bearing on whether they're blocked or not. This is intentional. Providing it allows us a way to access the various *"Other options and special flags"* without the need for the rule to take any more apparent action upon the request. However, "profiled" requests will be denoted as such on the IP test page when testing an IP, and can be also be chained to other rules if wanted.

Toggling off IP tracking on a new profile auxiliary rule, setting the conditions accordingly, should achieve the desired result here. :-)

**Comment by mikeruss1** [comment]
cool, thank you very much.

and your expectation is it wont change the number of times flat files are reset?

**Comment by Maikuolan** [comment]
> and your expectation is it wont change the number of times flat files are reset?

In theory, it's possible that it may marginally reduce the number of resets, but I doubt the difference will be meaningful enough to notice it.

**Comment by Maikuolan** [comment]
*Marking as resolved and closing.*
